src/tactic/norm_cast.lean

```diff
@@ -25,268 +45,582 @@ try_for 1000 (mk_instance e)
+private meta def count_internal_coes (e : expr) : ℕ :=
+count_coes e - count_head_coes e
+
+/-- aux function for `norm_cast.classify_type` -/
+private meta def classify_type_aux_old (lhs rhs : expr) : tactic label :=
+do
+  let lhs_head_coes := count_head_coes lhs,
+  let lhs_internal_coes := count_internal_coes lhs,
+  let rhs_head_coes := count_head_coes rhs,
+  let rhs_internal_coes := count_internal_coes rhs,
+  if lhs_head_coes = 0 ∧ lhs_internal_coes ≥ 1 then
+    return elim -- abs ↑n = ↑n
+    --if rhs_head_coes = 0 ∧ rhs_internal_coes = 0 then
+    --  return elim
+    --else
+    --  fail "norm_cast: badly shaped elim lemma, rhs can't contain coes"
+  else if lhs_head_coes = 1 ∧ lhs_internal_coes = 0 then
+    if rhs_head_coes = 0 then
+      if rhs_internal_coes ≥ 1 then
+        return move
+      else
+        return squash
+    else
+      fail "norm_cast: badly shaped lemma, rhs can't start with coe"
+  else if lhs_head_coes ≥ 2 ∧ lhs_internal_coes = 0 then
+    if rhs_head_coes ≤ lhs_head_coes ∧ rhs_internal_coes = 0 then
+      return squash
+    else
+      fail "norm_cast: badly shaped squash lemma"
+  else
+    fail "norm_cast: lhs must contain at least one coe"
+
+--old version, kept for debuging purposes
+/-- TODO: update and describe -/
+meta def classify_type_old (ty : expr) : tactic label :=
+do (args, tp) ← mk_local_pis ty,
+match tp with
+| `(%%lhs = %%rhs) := classify_type_aux_old lhs rhs
+| `(%%lhs ↔ %%rhs) := classify_type_aux_old lhs rhs
+| _ := fail "norm_cast: lemma must be = or ↔"
+end
 
 /--
-This is an attribute for simplification rules that are
-used to normalize casts.
+aux function for `norm_cast.classify_type`
 
-Let r be = or ↔, then elimination lemmas of the shape
-Π ..., P ↑a1 ... ↑an r P a1 ... an should be given the
-attribute elim_cast.
+remark: the classifier is a little bit less restrictive than the rules describing to the labels.
+This is a consequence of the fact that the rules apply to the explicit coercions appearing in the expression.
+'coe_to_sort' can tipically appear in type annotations that are implicit in the displayed expression, but
+will be accounted for in 'count_coes'.
+TODO: see 'count_coes'
 -/
-@[user_attribute]
-meta def elim_cast_attr : user_attribute simp_lemmas :=
-{
-    name      := `elim_cast,
-    descr     := "attribute for lemmas of the shape Π ..., P ↑a1 ... ↑an = P a1 ... an",
-    cache_cfg :=
-        { mk_cache     := mk_cache,
-          dependencies := [], },
-}
+private meta def classify_type_aux (lhs rhs : expr) : tactic label :=
+do
+  if count_coes lhs = 0 then
+    fail "norm_cast: badly shaped lemma, lhs must contain at least one coe"
+  else skip,
+  let lhs_head_coes     := count_head_coes lhs,
+  let lhs_internal_coes := count_internal_coes lhs,
+  let rhs_head_coes     := count_head_coes rhs,
+  let rhs_internal_coes := count_internal_coes rhs,
+  if lhs_head_coes = 0 then do
+    return elim
```

Comment:
An `elim` lemma can't have any casts on the rhs, right? Can we test for this so we have better failure behavior?

Reply:
the risk is that there's an implicit 'coe_to_sort' somewhere on the right.
I intended to add guards once the 'count_coe' functions is fixed (i.e. once it doesn't count implicit coercions anymore)